Resolving fcrontab warnings when run as root

yo8192 opened the pull request:
(raising on behalf of tux88)

I have encountered a problem for a long time with pam and nslcd,
which produces an unnecessary warning when calling
  "fcrontab <crontabfile>"
as user root.
It produces an output similar in /var/log/warn
  "nslcd[673]: [...] <passwd=""> request denied by validnames option"

yo8192 commented on fileconf.c:
```diff
@@ -1279,7 +1279,7 @@ check_username(char *ptr, cf_t * cf, cl_t * cl)
     username[indx] = '\0';
 
-    if ((userpwent = getpwnam(username)) != NULL) {
+    if (username[0]!='\0' && (userpwent=getpwnam(username))!=NULL) {
```
It would be best to split this line: 1 if statement line to check if username  is empty, one to set userpwent, and one to check if userpwent!=NULL.

Also small style comment, but the code usually has spaces around '!=' so best match this for consistency.

tux88 replied:
all crontab-Files for user=root generate the error, here is an example:
==========
!erroronlymail(true)
!mailto(root)
IN_BOOT=
IN_RUNLEVEL=
PATH=/bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin
20  2      * * *   root  /usr/local/sbin/....
==========
/usr/bin/fcrontab <crontab-file>
==========

I think the style (one/two lines, spaces) are cosmetic, here are another choices:
==========
if (username[0] == '\0') { }
else if ((userpwent = getpwnam(username)) != NULL) {
==========
if (username[0] != '\0') { 
  userpwent = getpwnam(username);
  if (userpwent != NULL) {
==========

With the first solution, the original query does not have to be changed, the second solution is smart for beginners, but I find my original code (above) better and more compact (c is short and sweet).

yo8192 replied:
In terms of code style, in the early days of fcron I used a lot of "compact" lines with variable assignment and test on a single line.  They are still plenty of examples in the code.

However since then, I have come to value code clarity over compact code, and prefer splitting over multiple lines for this reason.  Another benefit is that it tends to make future diffs (when the code will change later) clearer / easier.

Based on this I would favour the second solution.

yo8192 commented:
Hi tux88,

Apologies for the delay.
I've tried reproducing your issue, without success (when using a valid config line).

It looks like the fcrontab file example you proposed had some formatting issue, I assume when copying as a comment in github.  I have fixed the formatting and used:
```
!erroronlymail(true)
!mailto(root)
IN_BOOT=
IN_RUNLEVEL=
PATH=/bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin
20 2 * * *       root /bin/true
```
Can you try with this exact example and tell me if you still see the warning?

Note that in my case I don't have nslcd, but  instead I've added a debug statement to print the value of 'username' right before the if statement you are proposing to change:

```
/FIXME TEMPORARY DEBUG
    fprintf(stderr, "username='%s'\n", username);
    if ((userpwent = getpwnam(username)) != NULL) {
```

Based on your code change proposal, it seems that the issue would be if `username=''`, but I could not reproduce this with a valid config line.

It seems to me the code before this function `check_username()` is called should skip all blank spaces before calling `check_username()`, so I can't see how `username` could be empty, unless:
* the word being checked starts with a non-space, non-alpha-numerical character (e.g. `%word`).  I am not clear how that could happen for a valid line, but could be missing something?
* the line stops right after the 'day-of-week' field, i.e. no command was specified (which is invalid and fcron will flag it)

e.g. each of the two lines below will get me `username=''`
```
20 2 * * * %root /bin/true
20 2 * * * 
```

I'd be quite curious to confirm which line does cause the issue on your side, so I can reproduce here

yo8192 commented:
closing this old PR... feel free to reopen / open a new one answering the remaining questions and addressing the points I raised if you're still interested in this.
